godot-core/src/builtin/basis.rs

```diff
@@ -0,0 +1,799 @@
+            Basis::from_rows_array(&[b.cos(), 0.0, b.sin(), 0.0, 1.0, 0.0, -b.sin(), 0.0, b.cos()]);
+        let zmat =
+            Basis::from_rows_array(&[c.cos(), -c.sin(), 0.0, c.sin(), c.cos(), 0.0, 0.0, 0.0, 1.0]);
+
+        match order {
+            EulerOrder::XYZ => xmat * ymat * zmat,
+            EulerOrder::XZY => xmat * zmat * ymat,
+            EulerOrder::YXZ => ymat * xmat * zmat,
+            EulerOrder::YZX => ymat * zmat * xmat,
+            EulerOrder::ZXY => zmat * xmat * ymat,
+            EulerOrder::ZYX => zmat * ymat * xmat,
+        }
+    }
+
+    /// Creates a `Basis` with a rotation such that the forward axis (-Z) points
+    /// towards the `target` position.
+    ///
+    /// The up axis (+Y) points as close to the `up` vector as possible while
+    /// staying perpendicular to the forward axis. The resulting Basis is
+    /// orthonormalized. The `target` and `up` vectors cannot be zero, and
+    /// cannot be parallel to each other.
+    ///
+    /// _Godot equivalent: `Basis.looking_at()`_
+    pub fn new_looking_at(target: Vector3, up: Vector3) -> Self {
+        super::inner::InnerBasis::looking_at(target, up)
+    }
+
+    /// Creates a `[Vector3; 3]` with the columns of the `Basis`.
+    pub fn to_cols(self) -> [Vector3; 3] {
+        self.transposed().rows
+    }
+
+    /// Creates a [`Quaternion`] representing the same rotation as this basis.
+    ///
+    /// _Godot equivalent: `Basis()`_
+    pub fn to_quat(self) -> Quaternion {
+        glam::Quat::from_mat3(&self.orthonormalized().to_glam()).to_front()
+    }
+
+    const fn to_rows_array(self) -> [f32; 9] {
+        let [Vector3 {
+            x: ax,
+            y: bx,
+            z: cx,
+        }, Vector3 {
+            x: ay,
+            y: by,
+            z: cy,
+        }, Vector3 {
+            x: az,
+            y: bz,
+            z: cz,
+        }] = self.rows;
+        [ax, bx, cx, ay, by, cy, az, bz, cz]
+    }
+
+    /// Returns the scale of the matrix.
+    ///
+    /// _Godot equivalent: `Basis.get_scale()`_
+    #[must_use]
+    pub fn scale(&self) -> Vector3 {
+        let det = self.determinant();
+        let det_sign = if det < 0.0 { -1.0 } else { 1.0 };
+
+        Vector3::new(
+            self.col_a().length(),
+            self.col_b().length(),
+            self.col_c().length(),
+        ) * det_sign
+    }
+
+    /// Returns the rotation of the matrix in euler angles.
+    ///
+    /// The order of the angles are given by `order`.
+    ///
+    /// _Godot equivalent: `Basis.get_euler()`_
+    pub fn to_euler(self, order: EulerOrder) -> Vector3 {
+        use glam::swizzles::Vec3Swizzles as _;
+        use glam::Vec3;
+
+        let col_a = self.col_a().to_glam();
+        let col_b = self.col_b().to_glam();
+        let col_c = self.col_c().to_glam();
+
+        let row_a = self.rows[0].to_glam();
+        let row_b = self.rows[1].to_glam();
+        let row_c = self.rows[2].to_glam();
+
+        match order {
```

Comment:
I assume the the remaining two branches `XYZ` and `YXZ` cannot be integrated into `to_euler_inner()`? I see the issues with the code, but from a conceptual level, shouldn't those be somewhat symmetric?

Otherwise it should at least be possible to make a different function that handles `XYZ` and `YXZ`.